### ISSUE TITLE
ESMF 8.3.0 and GCHP 14.1.0: PIOc errors at linking

### ISSUE THREAD

**Issue by sdeastham** [geoschem/gchp#305]
Building GCHP 14.1.1 with ESMF 8.3.0, I receive an error during the final linking step for GCHP:

```
[100%] Building Fortran object src/CMakeFiles/gchp.dir/GCHPctm.F90.o
[100%] Linking Fortran executable ../bin/gchp
/home/seastham/ESMF/ESMF_v830/INSTALL_gcc1130_omp4_ucx_hybrid_ESMF830/lib/libO/Linux.gfortran.64.openmpi.default/libesmf.a(ESMCI_Mesh_FileIO.o): In function `ESMCI_mesh_create_from_ESMFMesh_file(int, char*, bool, ESMC_CoordSys_Flag, ESMCI::DistGrid*, ESMCI::Mesh**)':                                                                                                                                                                                                               ESMCI_Mesh_FileIO.C:(.text+0x754): undefined reference to `PIOc_Set_IOSystem_Error_Handling'
ESMCI_Mesh_FileIO.C:(.text+0x773): undefined reference to `PIOc_openfile'
ESMCI_Mesh_FileIO.C:(.text+0x7ce): undefined reference to `PIOc_openfile'
ESMCI_Mesh_FileIO.C:(.text+0x11de): undefined reference to `PIOc_closefile'
/home/seastham/ESMF/ESMF_v830/INSTALL_gcc1130_omp4_ucx_hybrid_ESMF830/lib/libO/Linux.gfortran.64.openmpi.default/libesmf.a(ESMCI_Mesh_FileIO.o): In function `ESMCI_mesh_create_from_UGRID_file(int, char*, bool, ESMC_CoordSys_Flag, ESMC_MeshLoc_Flag, char*, ESMCI::DistGrid*, ESMCI::Mesh**)':
ESMCI_Mesh_FileIO.C:(.text+0x19c3): undefined reference to `PIOc_Set_IOSystem_Error_Handling'
ESMCI_Mesh_FileIO.C:(.text+0x19e2): undefined reference to `PIOc_openfile'
ESMCI_Mesh_FileIO.C:(.text+0x1a3d): undefined reference to `PIOc_openfile'
ESMCI_Mesh_FileIO.C:(.text+0x24e0): undefined reference to `PIOc_closefile'
/home/seastham/ESMF/ESMF_v830/INSTALL_gcc1130_omp4_ucx_hybrid_ESMF830/lib/libO/Linux.gfortran.64.openmpi.default/libesmf.a(ESMCI_Mesh_FileIO.o): In function `ESMCI_mesh_create_from_file(char*, ESMC_FileFormat_Flag, bool, bool, ESMC_CoordSys_Flag, ESMC_MeshLoc_Flag, char*, ESMCI::DistGrid*, ESMCI::DistGrid*, ESMCI::Mesh**, int*)':
ESMCI_Mesh_FileIO.C:(.text+0x2a78): undefined reference to `PIOc_Init_Intracomm'
ESMCI_Mesh_FileIO.C:(.text+0x2c71): undefined reference to `PIOc_free_iosystem'
```

These errors continue for a while. Compiler was GNU 11.3.0 with OpenMPI 4.1.5.

Environment variables set during build of ESMF:
```ESMF_COMM=openmpi
ESMF_COMPILER=gfortran
ESMF_OPENMP=ON
ESMF_NETCDF=nc-config
ESMF_PIO=internal
```

This is being built on the MIT Svante cluster.

**Comment by lizziel** [comment]
This may have something to do with this update in [ESMF 8.3.0](https://github.com/esmf-org/esmf/releases/tag/v8.3.0):

> ESMF uses a library called [ParallelIO (PIO)](https://github.com/NCAR/ParallelIO) for its internal I/O operations, such as reading in mesh files and writing out fields. During this release, the version of PIO used internally was upgraded from a very outdated 1.x version to version 2.5. A new option was also added to the ESMF build system to allow linking to an external build of the PIO library.

**Comment by sdeastham** [comment]
I strongly suspect you're right @lizziel - unfortunately I have no idea quite what to do about it. I had wondered if something in MAPL was referring to things that no longer exist in the new PIO, but these seem to be calls coming from ESMF itself.. also, I should mention that building with ESMF 8.4.1 does not appear to resolve this issue if I recall correctly (but happy to try again if it's thought that a fix was implemented).

**Comment by lizziel** [comment]
I'm digging through the code now, both ESMF and various GMAO repos. @mathomp4 might beat me to it. Matt, do you use internal or external PIO? One possibility is if you use external and therefore have not run into this issue.

**Comment by lizziel** [comment]
Have you tried excluding PIO? I don't see anything in either the MAPL or ESMF release notes that specifying it became a requirement since 8.1.

**Comment by mathomp4** [comment]
@lizziel We do not externally build ParallelIO when we build ESMF. We use the built-in PIO. The error above looks like the "you now need to add `-lpioc` to your link line". 

In our build system we are using the ["official" `FindESMF.cmake`](https://github.com/esmf-org/esmf/blob/develop/cmake/FindESMF.cmake) from ESMF. It works off of ESMFMKFILE and since ESMF makes that, `-lpioc` comes with it.

Your [`FindESMF.cmake` file](https://github.com/geoschem/GCHP/blob/main/cmake/FindESMF.cmake) is different. If I had to guess, maybe you need to add `pioc` to:

https://github.com/geoschem/GCHP/blob/0345abf5cb237a72e0fc33873263f6618ff3e16b/cmake/FindESMF.cmake#L138

I can say if I look at my `ESMF.mk`, I see `-lpioc` at the end of `ESMF_F90LINKLIBS` and `ESMF_F90ESMFLINKLIBS`

**Comment by lizziel** [comment]
I'll try this. @sdeastham, I built my own ESMF so no need to test this right away. I'll report back on what I find.

**Comment by sdeastham** [comment]
Thanks @mathomp4 and @lizziel ! This is consistent with something I just tried - specifically, adding the path to `libpioc.a` (within the ESMF `lib` directory) to the link command. This seems to have fixed the issue, as GCHP built successfully and a test run at C24 has at least made it through the first couple of time steps. It sounds like @mathomp4 's fix is much more permanent than my kludge though.

**Comment by lizziel** [comment]
Thanks @mathomp4. I'll try to get that working. I'm not sure the history of why we have a custom FindESMF.cmake.

Incidentally, I also tried building 8.3 without setting ESMF_PIO. GCHP built and ran successfully using that install, so it does not appear to be a required setting. @mathomp4, do you know if there is a significant performance boost from using it? I'll try to assess this myself too.

**Comment by mathomp4** [comment]
@lizziel I don't know. You could ask the ESMF folks. I just build a pretty stock ESMF and figure if they turned it on, I'd use it. 

That said, I don't know if GEOS/MAPL uses any of the ESMF write routines anyway so it's possible we'd never see it!

**Comment by sdeastham** [comment]
One bit of additional information: if I look at `CMakeCache.txt` in the build directory (for an unmodified GCHP clone), I see that the pioc libraries _are_ being detected by our `FindESMF.cmake` routine. Unfortunately I am very limited when it comes to cmake so I'm not at all sure why that information isn't then making it to the final link step. I've attached my `CMakeCache.txt` for reference; if you search for `pioc` you should see that it is indeed identifying both `-lpioc` as a necessary linker argument and adding the path to the `libpioc.a` file to `ESMF_3RD_PARTY_LINKLIB_PIOC`. Looking at our `FindESMF.cmake` it then seems like this _should_ result in that path being added to `ESMF_LIBRARIES` - but at that point I'm not sure what happens. Either:
* It's not being added for some reason; or
* It is added, but `ESMF_LIBRARIES` isn't used later; or
* `ESMF_LIBRARIES` is used later, but whatever it is used in doesn't make it to the final link step

Obviously I'm a bit out of my depth here..

[CMakeCache.txt](https://github.com/geoschem/GCHP/files/11133900/CMakeCache.txt)

**Comment by sdeastham** [comment]
FYI - it looks like adding `ESMF_LIBRARIES` to `FindESMF.cmake` (so that it reads as follows) might fix the issue:

```
# Make an imported target for ESMF
if(NOT TARGET esmf)
    add_library(esmf STATIC IMPORTED GLOBAL)
    set_target_properties(esmf PROPERTIES
        IMPORTED_LOCATION ${ESMF_LIBRARY}
    )
    target_link_libraries(esmf
        INTERFACE
            ${ESMF_LIBRARIES}
            ${NETCDF_LIBRARIES}
            ${MPI_Fortran_LIBRARIES} ${MPI_CXX_LIBRARIES}
            rt ${stdcxx} ${libgcc}
    )
    target_include_directories(esmf INTERFACE ${ESMF_INCLUDE_DIRS})
    add_library(ESMF ALIAS esmf)
endif()
```

**Comment by lizziel** [comment]
I'm not convinced that we even need to build ESMF with PIO. It appears the MAPL I/O subroutines that do I/O are defined in MAPL rather than ESMF, although they are wrapped by ESMF_* interfaces. @tclune, could you weigh in on if MAPL uses ESMF PIO?

**Comment by tclune** [comment]
MAPL does *not* use PIO.  We do have some wrappers that maybe have names that look a bit like ESMF names,  but even that should be rare and are probably not even related to the sort of I/O we are talking about here.

**Comment by sdeastham** [comment]
My concern is more that a user with a valid ESMF build will now see an error, fundamentally because our FindESMF is missing a linker variable. I think the fix suggested above is benign for non-PIO builds?

**Comment by lizziel** [comment]
That's a very valid point.

**Comment by tclune** [comment]
@mathomp4  Do you have any guidance for what might go wrong and how we'd deal with it?

**Comment by mathomp4** [comment]
I think the proposal by @sdeastham could work. It's hard reading find_package files, but I think that would work. I keep thinking it's duplicating something, but then again, if it works, that's good enough.

My main advice would be to eventually move to the "official" FindESMF.cmake package. Of course, even then we have to do weird stuff because of GEOS historical-ness:
```cmake
    find_package(ESMF 8.4.0 MODULE REQUIRED)

    # ESMF as used in MAPL requires MPI
    # NOTE: This looks odd because some versions of FindESMF.cmake out in the
    #       world provide an "esmf" target while others provide "ESMF". So we
    #       need this ugliness to support both.
    if (TARGET esmf)
      target_link_libraries(esmf INTERFACE MPI::MPI_Fortran)
    else()
      target_link_libraries(ESMF INTERFACE MPI::MPI_Fortran)
      # MAPL and GEOS use lowercase target due to historical reasons but
      # the latest FindESMF.cmake file from ESMF produces an ESMF target.
      add_library(esmf ALIAS ESMF)
    endif()
```

**Comment by sdeastham** [comment]
I've now discovered that I also get an error on a completely separate system for a similar reason. I will make a recommendation which should simplify the file while also resolving both issues, but would propose that we move towards @mathomp4 's suggestion of using the official `FindESMF.cmake` in the long run.

**Comment by lizziel** [comment]
Thank you @sdeastham. I agree on using the official `FindESMF.cmake`.

**Comment by lizziel** [comment]
A fix for this will go into version 14.2.0. Future work will update to use the official FindESMF.cmake, or at least a version as close to the official as possible.
